File: 50SoG/SGDraw.m
//
//  SGDraw.m
//  50SoG
//
//  Created by Alexander v. Below on 15.09.15.
//  Copyright © 2015 AVB Software. All rights reserved.
//

#import "SGDraw.h"
#if TARGET_OS_IPHONE
#import <UIKit/UIKit.h>
#define OSColor UIColor
#else
#import <AppKit/AppKit.h>
#define OSColor NSColor
#endif

void drawShades(CGRect rect, CGRect bounds)
{
    CGContextRef ctx;
#if TARGET_OS_IPHONE
    ctx = UIGraphicsGetCurrentContext();
#else
    ctx = [NSGraphicsContext currentContext].CGContext;
#endif
    CGFloat fullWidth = bounds.size.width;
    CGFloat oneShadeWidth = fullWidth / 50.0;
    
    int shade;
    int firstShade = floor (rect.origin.x * 50.0 / fullWidth);
    int lastShade = floor (CGRectGetMaxX(rect) * 50.0 / fullWidth);
    for (shade = firstShade; shade <= lastShade; shade ++)
    {
        CGFloat grey = 1.0 - shade / 49.0;
        OSColor *shadeOfGrey = [OSColor colorWithWhite:grey alpha:1.0];
        [shadeOfGrey set];
        CGContextAddRect(ctx, CGRectMake(shade*oneShadeWidth, rect.origin.y, oneShadeWidth + 2, rect.size.height));
        CGContextFillPath(ctx);
    }
}

File: 50SoG/SGDraw.swift
//
//  SGDraw.swift
//  50SoG
//
//  Created by Alexander v. Below on 05.06.19.
//  Copyright © 2019 AVB Software. All rights reserved.
//

#if TARGET_OS_IPHONE
typealias OSColor = UIColor
#else
import UIKit
typealias OSColor = UIColor
#endif
@objc public class objcWrapper: NSObject {
    @objc static func objcDrawShades(rect: CGRect, bounds: CGRect) {
        drawShades(rect: rect, bounds: bounds)
    }
}

public func drawShades(rect: CGRect, bounds: CGRect) {
    var ctx : CGContext!
    ctx = UIGraphicsGetCurrentContext()
    ctx.saveGState()
    defer {
        ctx.restoreGState()
    }
    let fullWidth = bounds.size.width
    let oneShadeWidth = fullWidth / 50.0
    
    let firstShade = Int(floor (rect.origin.x * 50.0 / fullWidth))
    let lastShade = Int(floor (rect.maxX * 50.0 / fullWidth))
    for shade in firstShade...lastShade {
        
        let grey = 1.0 - CGFloat(shade) / 49.0
        let shadeOfGrey = UIColor(white: grey, alpha: 1.0)
        shadeOfGrey.setFill()
        let rect = CGRect(x: CGFloat(shade)*oneShadeWidth, y: rect.origin.y, width: oneShadeWidth + 2, height: rect.size.height)
        ctx.addRect(rect)
        ctx.fillPath()
    }
}
